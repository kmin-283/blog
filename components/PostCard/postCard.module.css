.postCard {
    position: relative;
    top: 0;
    width: 20rem;
    height: 22.5rem;
    box-shadow: 0 0 2px 1px var(--postcard-shadow-color);
    margin-top: 3rem;
    transition: top 0.3s;
}

.postCard:hover {
    top: -20px;
}

.postCard a {
    cursor: pointer;
}

.thumbnail {
    position: relative;
    height: 35%;
}

.about {
    position: absolute;
    color: black;
    top: 30%;
    left: 10%;
    right: 10%;
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    justify-content: flex-start;
    /*align-items: center;*/
    background-color: white;
    padding: 0.5rem;
    height: 70%;
    box-shadow: 0 -10px 10px 1px var(--postcard-shadow-color);
}

.title {
    font-size: 1.25rem;
    margin: 1rem 0 0.5rem 0;
}

.divider {
    margin: 0.125rem 0 0.375rem 0;
    border: 1px solid rgba(96, 96, 96, 0.1);
}

.description {
    flex-grow: 2;
    font-size: 0.875rem;
}

.subInfo {
    width: 100%;
}

.time {
    font-size: 0.75rem;
    color: var(--icon-inactive-color);
}
