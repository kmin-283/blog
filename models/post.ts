import mongoose, { model, Schema } from "mongoose";

export interface IPost {
  title: string;
  tags: string[];
  file: string;
}

const postSchema = new Schema<IPost>(
  {
    title: { type: String, required: true },
    tags: [String],
    file: { type: String, required: true },
  },
  {
    timestamps: true,
  }
);

export default mongoose.models.Post || model<IPost>("Post", postSchema);
