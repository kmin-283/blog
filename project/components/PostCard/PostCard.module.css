.postCard {
    position: relative;
    top: 0;
    width: 18rem;
    height: 32rem;
    box-shadow: 0 0 2px 1px var(--postcard-shadow-color);
    margin: 2rem 1rem;
    transition: top 0.3s, box-shadow 0.3s, padding 0.3s;
    border-radius: 5px;
}

.postCard:hover {
    top: -20px;
    padding: 5px;
    box-shadow: 0 0 4px 4px var(--main-theme-color);
    border-radius: 5px;
}

.postCard a {
    cursor: pointer;
}

.thumbnail {
    position: relative;
    height: 70%;
}

.about {
    position: absolute;
    color: black;
    bottom: 0;
    left: 0;
    right: 0;
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    justify-content: flex-start;
    background-color: white;
    padding: 0.5rem;
    height: 30%;
    box-shadow: 0 -10px 10px 1px var(--postcard-shadow-color);
}

.title {
    font-size: 1.25rem;
    line-height: 1.5rem;
    margin: 1rem 0 0.5rem 0;
}

.divider {
    margin: 0.125rem 0 0.375rem 0;
    border: 1px solid rgba(96, 96, 96, 0.1);
}

.description {
    flex-grow: 2;
    font-size: 0.875rem;
    text-overflow: ellipsis;
}

.subInfo {
    width: 100%;
}

.time {
    font-size: 0.75rem;
    color: var(--icon-inactive-color);
}
