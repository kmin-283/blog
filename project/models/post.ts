import { Schema } from "mongoose";

export interface IPost {
  _id: string;
  title: string;
  tags: string[];
  file: string;
  thumbnail: string;
  description: string;
  internalLinks?: string;
  updatedAt: Date;
  createdAt: Date;
}

export interface IPostWithMarkdown extends IPost {
  markdown: string;
}

const postSchema = new Schema<IPost>(
  {
    title: { type: String, required: true },
    tags: { type: [String] },
    file: { type: String, required: true },
    thumbnail: { type: String, required: true },
    description: { type: String, required: true },
    internalLinks: { type: String },
  },
  {
    timestamps: true,
  }
);

export default postSchema;
