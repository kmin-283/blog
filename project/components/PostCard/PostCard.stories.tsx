import React from "react";
import { Story, Meta } from "@storybook/react";
import PostCard from "@/components/PostCard/PostCard";
import { PostCardProps } from "@/components/PostCard/PostCard";

export default {
  title: "Components/PostCard",
  component: PostCard,
} as Meta;

const Template: Story<PostCardProps> = ({ post }) => {
  return <PostCard post={post} />;
};

export const BasicPostCard = Template.bind({});
BasicPostCard.args = {
  post: {
    _id: "1",
    title: "테스트 포스트 입니다",
    tags: ["test1", "test2"],
    file: "/__mocks__/fileMock.js",
    thumbnail: "/__mocks__/thumbnailMock.js",
    description: "postcard 컴포넌트 테스트",
    updatedAt: new Date("1995-12-17T03:24:00Z"),
    createdAt: new Date("1995-12-17T03:24:00Z"),
  },
};

export const LongTitlePostCard = Template.bind({});
LongTitlePostCard.args = {
  post: {
    _id: "1",
    title: "최적의 타이틀은 공백 포함 최대 32자 입니다ㅋㅋㅋㅋㅋㅋㅋ",
    tags: ["test1", "test2"],
    file: "/__mocks__/fileMock.js",
    thumbnail: "/__mocks__/thumbnailMock.js",
    description: "postcard 컴포넌트 테스트",
    updatedAt: new Date("1995-12-17T03:24:00Z"),
    createdAt: new Date("1995-12-17T03:24:00Z"),
  },
};

export const LongDescriptionPostCard = Template.bind({});
LongDescriptionPostCard.args = {
  post: {
    _id: "1",
    title: "최적의 타이틀은 공백 포함 최대 32자 입니다ㅋㅋㅋㅋㅋㅋㅋ",
    tags: ["test1", "test2"],
    file: "/__mocks__/fileMock.js",
    thumbnail: "/__mocks__/thumbnailMock.js",
    description:
      "이 포스트에 대한 설명은 길어야 합니다." +
      "왜냐하면 지정된 영역을 벗어났을 때 스타일링이 어떻게 구성되는지 확인하기 위함이죠" +
      "여기 포스트는 더 보이지 않을겁니다.",
    updatedAt: new Date("1995-12-17T03:24:00Z"),
    createdAt: new Date("1995-12-17T03:24:00Z"),
  },
};
