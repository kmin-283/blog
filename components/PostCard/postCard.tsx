import React, { FC } from "react";
import { IPost } from "../../models/post";
import Link from "next/link";
import Image from "next/image";
import { convertToKRDate } from "@/utils/time";
import styles from "./postCard.module.css";

interface PostCardProps {
  post: IPost;
}

const PostCard: FC<PostCardProps> = ({ post }) => {
  const time = convertToKRDate(post.updatedAt.toString());
  return (
    <div className={styles.postCard}>
      <Link href={`/${post.title.replace(/\s/g, "-")}`}>
        <a>
          <div className={styles.thumbnail}>
            {post.thumbnail && (
              <Image src={post.thumbnail} alt={"thumbnail"} layout={"fill"} />
            )}
          </div>
          <div className={styles.about}>
            <h2 className={styles.title} property={"title"}>
              {post.title}
            </h2>
            <hr className={styles.divider} />
            <span className={styles.description} property={"description"}>
              {post.description}
            </span>
            <div className={styles.subInfo}>
              <time className={styles.time} dateTime={time}>
                {time}
              </time>
            </div>
          </div>
        </a>
      </Link>
    </div>
  );
};

export default PostCard;
