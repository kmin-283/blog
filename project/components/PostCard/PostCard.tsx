import React, { FC } from "react";
import { IPost } from "@/models/post";
import Link from "next/link";
import Image from "next/image";
import { convertToKRDate } from "@/utils/time";
import styles from "./PostCard.module.css";

export interface PostCardProps {
  post: IPost;
}

const PostCard: FC<PostCardProps> = ({ post }) => {
  const createdTime = convertToKRDate(post.createdAt.toString());
  return (
    <Link href={`/${post.title.replace(/\s/g, "-")}`}>
      <a>
        <div className={styles.postCard}>
          <div className={styles.thumbnail}>
            {post.thumbnail && (
              <Image src={post.thumbnail} alt={"thumbnail"} layout={"fill"} />
            )}
          </div>
          <article className={styles.about}>
            <h2 className={styles.title} property={"title"}>
              {post.title}
            </h2>
            <hr className={styles.divider} />
            <span className={styles.description} property={"description"}>
              {post.description}
            </span>
            <time className={styles.time} dateTime={post.createdAt.toString()}>
              {createdTime}
            </time>
          </article>
        </div>
      </a>
    </Link>
  );
};

export default PostCard;
