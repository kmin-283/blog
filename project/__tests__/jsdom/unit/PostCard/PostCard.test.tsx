import React from "react";
import PostCard from "@/components/PostCard/PostCard";
import { cleanup, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { IPost } from "@/models/post";
import renderer from "react-test-renderer";

describe("postcard 컴포넌트", () => {
  const post: IPost = {
    _id: "1",
    title: "테스트 포스트 입니다",
    tags: ["test1", "test2"],
    file: "/__mocks__/fileMock.js",
    thumbnail: "/__mocks__/thumbnailMock.js",
    description: "postcard 컴포넌트 테스트",
    updatedAt: new Date("1995-12-17T03:24:00Z"),
  };
  afterAll(cleanup);

  test("Header 렌더링", () => {
    const postCard = renderer.create(<PostCard post={post} />).toJSON();
    expect(postCard).toMatchSnapshot();
  });

  test("title, thumbnail, description, time 정보가 있어야 한다", () => {
    const { getByRole, getByText } = render(<PostCard post={post} />);

    const link = getByRole("link");
    expect(link).toBeInTheDocument();

    const thumbnail = getByRole("img");
    expect(thumbnail).toBeInTheDocument();

    const title = getByText("테스트 포스트 입니다");
    expect(title).toBeInTheDocument();

    const description = getByText("postcard 컴포넌트 테스트");
    expect(description).toBeInTheDocument();

    const time = getByText("1995년 12월 17일");
    expect(time).toBeInTheDocument();
  });
});
