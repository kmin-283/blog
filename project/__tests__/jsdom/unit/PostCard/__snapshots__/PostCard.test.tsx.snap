// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcard 컴포넌트 Header 렌더링 1`] = `
<a
  href="/%ED%85%8C%EC%8A%A4%ED%8A%B8-%ED%8F%AC%EC%8A%A4%ED%8A%B8-%EC%9E%85%EB%8B%88%EB%8B%A4"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="postCard"
  >
    <div
      className="thumbnail"
    >
      <span
        style={
          Object {
            "background": "none",
            "border": 0,
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": "initial",
            "left": 0,
            "margin": 0,
            "opacity": 1,
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": "initial",
          }
        }
      >
        <img
          alt="thumbnail"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
        <noscript>
          <img
            alt="thumbnail"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            sizes="100vw"
            src="/_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=3840&q=75"
            srcSet="/_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=640&q=75 640w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=750&q=75 750w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=828&q=75 828w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=1080&q=75 1080w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=1200&q=75 1200w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=1920&q=75 1920w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=2048&q=75 2048w, /_next/image?url=%2F__mocks__%2FthumbnailMock.js&w=3840&q=75 3840w"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </noscript>
      </span>
    </div>
    <article
      className="about"
    >
      <h2
        className="title"
        property="title"
      >
        테스트 포스트 입니다
      </h2>
      <hr
        className="divider"
      />
      <span
        className="description"
        property="description"
      >
        postcard 컴포넌트 테스트
      </span>
      <time
        className="time"
        dateTime="Sun Dec 17 1995 12:24:00 GMT+0900 (대한민국 표준시)"
      >
        1995년 12월 17일
      </time>
    </article>
  </div>
</a>
`;
